File: Foundations of Aerodynamics Kuethe/airfoil.py
# Python 3
# Translation of FORTRAN program p. 133 from
# https://github.com/tristanCB/airfoil
# This program translates the FORTRAN program on p. 133 of 
# Foundations of Aerodynamics: Bases of Aerodynamic Design — Fourth Edition" from Kuethea. M. and Chowc-Y (1968)
# Author: Tristan CB
# Dependencies
import numpy as np
import math
# Summary of program
"""
This prgram computes velocity and pressure coefficients
around an airfoil, whose contor is approximated by vortex panels
of linearly varying strength.
Assumptions: 2D incompressible flow
"""
#
# Ideas
# Encoding equations into CNN would it be more efficient?
# Using it as a loss function in a ML model would be neat
#
# Will have to find ways of using CRAMER and DETERM subroutines ...

def determ(ARRAY, N, checkPrecision = False):
    """
    DETERM is the value of the determinant of an N*N matrix called ARRAY,
    computed by the technique of pivotal condensation. This function is 
    taken from p. 113 - 114 of Chow (1979)

    Notes on comparing it's precision and speed with numpy.
    """
    A = np.zeros((N,N))
    for i in range(N):
        for j in range(N):
            A[i,j] = ARRAY[i,j]
    M = 0
    while True:
        K = M + 1
        for i in range(K,N):
            RATIO = A[i,M]/A[M,M]
            for ij in range(K,N):
                A[i,ij] = A[i,ij] - A[M,ij]*RATIO
        if M == N-1:
            # Go to 4
            DETERM = 1
            for L in range(N):
                # print(f"A[L,L]: {A[L,L]}")
                DETERM = DETERM * A[L,L]
                # print(f"DETERM is now {DETERM}")
            # print(f"Returning {DETERM}")

            # Checks 
            if checkPrecision == True:
                deciNpDetC              = np.linalg.det(ARRAY)
                deciPivCondensationChow = DETERM
                diff = 1 - deciPivCondensationChow/deciNpDetC
                print(
                        f"""
                        Numpy func (np.linalg.det(ARRAY))   ==> {deciNpDetC}
                        Det. by pivitoal condensation       ==> {deciPivCondensationChow}
                        Diff                                  > {diff}
                        """)
                assert abs(diff) < 0.0001
                
            return DETERM
        else:
            M = M + 1
    pass

def cramer(C, A, X, N):
    """
    This is a funtion to emulate the subroutine cramer as used in the FORTRAN77 program
    Airfoil.
    This solves a set of algebraic equations
        C[i,j]*X[j] = A[i], i=1,2,3,...,N
    it was teken from p. 114 of Chow (1979)
    Note: Dimensions in the following statement must be equal to the numerical value of N.
    """
    CC = np.zeros((N,N))
    DENOM = determ(C, N)
    # np.set_printoptions(formatter={'float': '{: 0.6f}'.format})


    for k in range(N):
        for i in range(N):
            for j in range(N):
                CC[i,j] = C[i,j]
        for i in range(N):
            CC[i,k] = A[i]

        NUNOM = determ(CC, N)
        X[k] = NUNOM / DENOM
        # print(f"DENOM --> {DENOM}")
        # print(f"NUNOM --> {NUNOM}")
        # print(f"GAMMA will now be? --> {X[k]}")
    return X

def panelMethod(XB: np.array([]),YB: np.array([]), angle=8.0) -> dict:
    """
    # Specify coordinates (XB, YB) of boundary points on airfoil surface. 
    # The last point coinsides with the first
    XB = np.array([1.,.933,.750,.500,.250,.067,.0,.067,.25,.500,.750,.933, 1.0])
    YB = np.array([.0,-.005,-.017,-.033,-.042,-.033,.0,.045,.076,.072,.044,.013,0.])
    """
    # Get problem size depending on input data
    numberOfPanels = XB.shape[0]-1
    M = numberOfPanels
    MP1 = M + 1

    # Coordinates of control and boundary points
    # XB, YB  = (np.zeros((M)) for i in range(2))
    X, Y    = (np.zeros((M)) for i in range(2))
    # Represents γ' in text
    GAMMA   = np.zeros((M))

    S, SINE, COSINE, THETA, V, CP, RHS  = (np.zeros((M)) for i in range(7))
    CN1, CN2, CT1, CT2 =  (np.zeros((M,M)) for i in range(4))
    AN = np.zeros((MP1,MP1))

    AT = np.zeros((M,MP1))

    PI = 4.0 * np.arctan(1.0)
    ALPHA = angle * PI/180

    # Watch out the negative 1 was added befcause of zero based indexing??
    for i in range(M):
        IP1         = i + 1
        X[i]        = 0.5*(XB[i] + XB[IP1])
        Y[i]        = 0.5*(YB[i] + YB[IP1])
        S[i]        = np.sqrt( (XB[IP1] - XB[i])**2 + (YB[IP1] - YB[i])**2 )
        THETA[i]    = np.arctan2( (YB[IP1] - YB[i]), (XB[IP1] - XB[i]))
        SINE[i]     = np.sin(THETA[i])
        COSINE[i]   = np.cos(THETA[i])
        RHS[i]      = np.sin(THETA[i] - ALPHA)

    for i in range(M):
        for j in range(M):
            if i == j:
                # print(f"I is {i}, J is {j}") # UID for debugging
                CN1[i,j] = -1.0
                CN2[i,j] = 1.0
                CT1[i,j] = 0.5*PI
                CT2[i,j] = 0.5*PI
                # continue # break # still unsure about this
            else:
                A = (-(X[i] - XB[j]))*COSINE[j] - (Y[i]-YB[j])*SINE[j]
                B = (X[i]-XB[j])**2 + (Y[i]-YB[j])**2
                # B = ((X[i]-XB[j])**2) + ((Y[i] + YB[j])**2) # Watch out
                C = np.sin(THETA[i] - THETA[j])
                D = np.cos(THETA[i] - THETA[j])
                E = (X[i] - XB[j]) * SINE[j] - (Y[i] - YB[j])*COSINE[j]
                # print(f"ID: {i+1} {j+1}")
                F = np.log(1.0 + S[j]*(S[j] + 2.0 * A)/B)
                if np.isnan(F):
                    F = np.nan_to_num(F)
                G = np.arctan2(E*S[j], B+A*S[j])
                P = ( (X[i] - XB[j]) * np.sin(THETA[i] - 2*THETA[j]) 
                    + (Y[i] - YB[j]) * np.cos(THETA[i] - 2*THETA[j]) ) 
                Q = ( (X[i] - XB[j]) * np.cos(THETA[i] - 2*THETA[j]) 
                    - (Y[i] - YB[j]) * np.sin(THETA[i] - 2*THETA[j]) )
                CN2[i,j] = D + (0.5*Q*F/S[j]) - ((A*C + D*E)*G/S[j])
                CN1[i,j] = 0.5*D*F + C*G - CN2[i,j]
                CT2[i,j] = C + 0.5*P*F/S[j] + ((A*D-C*E)*G)/S[j]
                CT1[i,j] = ((0.5*C*F) - (D*G) - (CT2[i,j]))
                # For debugging
                # print(f"A, B, C, D, E, F, G, P, Q =====> {A}, {B}, {C}, {D}, {E}, {F}, {G}, {P}, {Q}")
                # print("CN2[i,j] ------>", CN2[i,j])
                # print("CN1[i,j] ------>", CN1[i,j])
                # print("CT2[i,j] ------>", CT2[i,j])
                # print("CT1[i,j] ------>", CT1[i,j])
    # print(A, B, C, D, E, F, G, P, Q)

    ##################
    # Coords (X,Y) of control point and panel length s are computed
    # for each of the vortex panels. RHS represents the right-hand side of eq. 47 p. 130.

    # Compute influence coeffs. in eqs. 47 and 49. respect.
    for i in range(M):
        AN[i,0] = CN1[i,0]
        AN[i,M] = CN2[i,M-1]
        AT[i,0] = CT1[i,0]
        AT[i,M] = CT2[i,M-1]
        for j in range(1,M):
            AN[i,j] = CN1[i,j] + CN2[i,j-1]
            AT[i,j] = CT1[i,j] + CT2[i,j-1]
    AN[M,0] = 1
    AN[M,M] = 1
    for j in range(1,M):
        AN[M,j] = 0
    RHS = np.pad(RHS, (0, 1), 'constant')

    # Solve eq. 47 for dimensionless strengths gamma using cramer's rule. 
    # Then compute and print dimensionless velocity and pressure coeefffs 
    # at control points.
    GAMMA = np.pad(GAMMA, (0, 1), 'constant')
    cramer(AN, RHS, GAMMA, MP1)
    for i in range(M):
        V[i] = np.cos(THETA[i] - ALPHA)
        for j in range(MP1):
            V[i] = V[i] + AT[i,j] * GAMMA[j]
            CP[i] = 1 - V[i]**2


    results = {}
    I = [i for i in range(M)]
    variables           = ["I","X", "Y","THETA", "S","GAMMA", "V", "CP"]
    corespondingArr  = [I, X, Y, THETA, S, GAMMA, V, CP]
    for i, ij in zip(variables,corespondingArr):
        results[i] = ij
    #     print(i)
    #     print(ij)
    return results

if __name__ == "__main__":
    # Uses matplotlib to plot Cp over and under airfoil
    import argparse
    import matplotlib.pyplot as plt
    from utils import rotateNumpy, plotFoil

    parser = argparse.ArgumentParser()
    
    parser.add_argument('--alpha',      type=float, help='Angle of attack. Float.', 
                        default=8.0)
    
    parser.add_argument('--NACA4Digit', type=int,   help='NACA 4 digits series number to be plotted', 
                        default=2412)
	
    args = parser.parse_args()

    # Parse user inputs
    angle           = args.alpha
    seriesNumber    = args.NACA4Digit
    # Show Foil  -- Recreation of fig frosm Kuethea (1968)

    # Numbers of panels
    panels = [12,48]
    plotFoil(seriesNumber,angle,panels=panels).show()
    plt.clf()

    # END #
    print(f"Done plotting NACA {seriesNumber}.")
    exit()

File: Foundations_of_Aerodynamics_Kuethe/airfoil.f
        PROGRAM AIRFOIL

        PARAMETER ( M = 12 )
        DIMENSION XB(M+1),YB(M+1),X(M),Y(M),S(M),SINE(M),COSINE(M), 
     * THETA(M) ,V(M) ,CP(M), GAMA(M+1) ,RHS(M+1) ,CN1(M,M),
     * CN2(M,M), CT1(M,M) ,CT2(M,M) ,AN(M+1,M+1) ,AT(M,M+1),
     * CN11(M)

        DATA XB 
     * /1.,.933,.750,.500,.250,.067,.0,.067,.25,.500,.750,.933, 1.0/
        DATA YB 
     * /.0,-.005,-.017,-.033,-.042,-.033,.0,.045,.076,.072,.044,.013,0./

        MP1 = M+1
        PI = 4.0 * ATAN(1.0)
        ALPHA =  8. * PI/180.

! C COORDINATES (X, Y) OF CONTROL POIN'!' PANEL LENGTH S ARE
! C COMPUTED FOR EACH OF THE VORTEX PANEL,. RHS REPRESEN'l'S
! C THE RIGHT-HAND SIDE OF EQ. (5.47).
        DO I = 1, M
        IP1 = I + 1
        X(I) = 0.5*(XB(I)+XB(IP1))
        Y(I) = 0.5*(YB(I)+YB(IP1))
        S(I) = SQRT( (XB(IP1)-XB(I))**2 + (YB(IP1)-YB(I))**2)
        THETA(I) = ATAN2( (YB(IP1)-YB(I)), (XB(IP1)-XB(I)) )
        SINE(I) = SIN( THETA (I) )
        COSINE(I) = COS( THETA (I) )
        RHS(I) = SIN( THETA(I)-ALPHA )
        END DO
        DO I = 1, M
        DO J = 1, M
        IF ( I .EQ. J ) THEN
        CN1(I,J) = -1.0
        CN2(I,J) = 1.0
        CT1(I,J) = 0.5*PI
        CT2(I,J) = 0.5*PI
        ELSE
        A = -(X(I)-XB(J))*COSINE(J) - (Y(I)-YB(J))*SINE(J)
        B = (X(I)-XB(J))**2 + (Y(I)-YB(J))**2
        C = SIN( THETA(I)-THETA(J) )
        D = COS ( THETA(I)-THETA(J) )
        E = (X(I)-XB(J))*SINE(J) - (Y(I)-YB(J))*COSINE(J)
        floatInp = (1. + S(J)*(S(J)+2.*A)/B)
        F = ALOG( 1.0 + S(J)*(S(J)+2.*A)/B )
        G = ATAN2( E*S(J), B+A*S(J) )
        P = (X(I)-XB(J)) * SIN( THETA(I)-2.*THETA(J) ) 
     *    + (Y(I)-YB(J)) * COS( THETA(I)-2.*THETA(J) )
        Q = (X(I)-XB(J)) * COS( THETA(I)-2.*THETA(J) ) 
     *    - (Y(I)-YB(J)) * SIN( THETA(I)-2.*THETA(J) )
        CN2(I,J) = D + .5*Q*F/S(J) - (A*C+D*E)*G/S(J)
        CN1(I,J) = .5*D*F + C*G - CN2(I,J)
        CT2(I,J) = C + .5*P*F/S(J) + (A*D-C*E)*G/S(J)
        CT1(I,J) = .5*C*F - D*G - CT2(I,J)
        ! For debugging purposes
        ! write(*, *) 'ID: ', I, J
        ! WRITE( *, * ) A, B, C, D, E, F, G, P, Q
        ! write(*, *) 'CN2 result: ', CN2(I,J) 
        ! write(*, *) 'CN1 result: ', CN1(I,J)
        ! write(*, *) 'CT2 result: ', CT2(I,J) 
        ! write(*, *) 'CT1 result: ', CT1(I,J) 
        END IF
        END DO
        END DO
C COMPUTE INFLUENCE COEFFICIENTS IN EQS.(s.47) AND (5.49),
C RESPECTIVELY.
        DO I = 1, M
        AN(I,1)     = CN1(I,1)
        AN(I,MP1)   = CN2(I,M)
        AT(I,1)     = CT1(I,1)
        AT(I,MP1)   = CT2(I,M)
            DO J = 2, M
            AN(I,J) = CN1(I,J) + CN2(I,J-1)
            AT(I,J) = CT1(I,J) + CT2(I,J-1)
            END DO
        END DO
        AN(MP1,1) = 1.0
        AN(MP1,MP1) = 1.0
        DO J = 2, M
        AN(MP1,J) = 0.0
        END DO
        RHS(MP1) = 0.0
C SOLVE EQ. (5.47) FOR DIMENSIONLESS STRENGTHS GAMA USING
C CRAMER'S RULE. THEN COMPUTE AND PRINT DIMENSIONLESS
C VELOCITY AND PRESSURE COEFFICIENT AT CONTROL POINTS.
        WRITE (6,6)
6       FORMAT(1H1///11X,1HI,4X,4HX(I),4X,4HY(I),4X,8HTHETA(I),
     *               3X,4HS(I),3X,7HGAMA(I),3X,4HV(I),6X,5HCP(I)/
     *               10X,3H---,3X,4H----,4X,4H----,4X,8H--------,
     *               3X,4H----,3X,7H-------,3X,4H----,6X,5H-----)
        CALL CRAMER ( AN, RHS, GAMA, MP1 )
        DO I = 1, M
        V(I) = COS( THETA(I)-ALPHA )
            DO J = 1, MP1
            V(I) = V(I) + AT(I,J)*GAMA(J)
            CP(I) = 1.0 - V(I)**2
            END DO
        WRITE(6,9) I,X(I),Y(I),THETA(I),S(I),GAMA(I),V(I),CP(I)
        END DO
9       FORMAT(10X,I2,F8.4,F9.4,F10.4,F8.4,2F9.4,F10.4)
        WRITE(6,10) MP1,GAMA(MP1)
10      FORMAT(10X,I2,35X,F9.4)



        STOP
        END

        SUBROUTINE CRAMER( C, A, X, N )
C THIS SUBROUTINE SOLVES A SET OF ALGEBRAIC EQUATIONS
C C(I,J)*X(J) = A(I), I=1,2,---,N
C IT IS TAKEN FROM P.114 OF CHOW(1979)
        PARAMETER ( M = 12 )
        DIMENSION C(M+1,M+1),CC(M+1,M+1),A(M+1),X(M+1)
        DENOM = DETERM( C, N )
        write (*,*), C
        DO K = 1, N
            DO I = 1, N
                DO J = 1, N
                CC(I,J) = C(I,J)
                END DO
            END DO
            DO I = 1, N
            CC(I,K) = A(I)
            END DO
        PUNOM = DETERM( CC, N )
        X(K) = DETERM( CC, N ) / DENOM
        ! Debugging
        ! write (*, *) "Denominator", DENOM
        ! write (*, *) "Numerator", PUNOM
        ! write (*, *) "GAMMA", X(K)

        END DO
        RETURN
        END
        FUNCTION DETERM ( ARRAY, N )
C DETERM IS THE VALUE OF THE DETERMINANT OF AN N*N
C MATRIX CALLED ARRAY, COMPUTED BY THE TECHNIQUE
C OF PIVOTAL CONDENSATION. THIS FUNCTION IS TAKEN
C FROM PP.113-114 OF CHOW(1979)
        PARAMETER ( M = 12 )
        DIMENSION ARRAY(M+1,M+1),A(M+1,M+1)
        DO I = 1, N
        DO J = 1, N
        A(I,J) = ARRAY(I,J)
        END DO
        END DO
        L = 1
1       K = L + 1
        DO I = K, N
        RATIO = A(I,L)/A(L,L)
            DO J = K, N
            A(I,J) = A(I,J) - A(L,J)*RATIO
            END DO
        END DO
        L = L + 1
        IF( L .LT. N ) GO TO 1
        DETERM =1.
        DO L = 1, N
        DETERM = DETERM * A(L,L)
        END DO
        RETURN
        END
